libs/appstore/src/appstore/appstore-app-client.ts

```diff
@@ -140,6 +147,235 @@ export class AppstoreAppClient {
+        //   throw new Error('Hash of received WebappEntry does not match the requested hash.');
+        // }
+
+        // 2.2 get UI bytes
+        // 2.2.1 get UIEntry
+        // TODO fetch the uiEntry first to check for the file size
+        // const uiEntryEntity = await this.portalZomeClient.customRemoteCall<Entity<UiEntry>>({
+        //   host,
+        //   call: {
+        //     dna: devhubDnaHash,
+        //     zome: 'apphub_csr',
+        //     function: 'get_ui_entry',
+        //     payload: webappEntryEntity.content.manifest.ui.ui_entry,
+        //   },
+        // });
+        // // 2.2.2 get UI bytes
+        // const uiAsset = await this.portalZomeClient.customRemoteCall<UiAsset>({
+        //   host,
+        //   call: {
+        //     dna: devhubDnaHash,
+        //     zome: 'apphub_csr',
+        //     function: 'get_ui_asset',
+        //     payload: webappEntryEntity.content.manifest.ui.ui_entry,
+        //   },
+        // });
+        // 2.3 get happ bundle
+        // 2.3.1 get AppEntry
+        // 2.3.2 For each role in the manifest, get dna bundle
+        // 2.3.2.1 get DnaEntry
+        // 2.3.2.2 for zome manifest in integrity and coordinator zomes get Wasm bytes
+        // 2.3.2.3 Combine all wasm bytes into bundle via Bundle.createDna() and then call .toBytes()
+        // 2.3.3 Create happ bundle with Bundle.createHapp and then .toBytes()
+        // 2.4 Create webhapp bundle with Bundle.createWebhapp and then call toBytes()
+        // 2.5 verify all hashes with the expected hashes
+
+        // return Promise.reject('fetchWebappBytes() is not implemented yet.');
+      },
+      dnaZomeFunction: {
+        dna: appVersion.apphub_hrl.dna,
+        zome: 'apphub_csr',
+        function: 'get_webapp_package_version_entry', // We just pick one of the functions for the sake of simplicity and assume that all other functions are callable as well by the same host
+      },
+      pingTimeout: 4000,
+    });
+  }
+
+  /**
+   *
+   * @param appVersion
+   * @param webappPackageVersion Can be provided if the WebAppPackageVersionEntry has already been fetched
+   * otherwise in order to omit redundant remote calls
+   * @returns
+   */
+  async fetchHappBytes(
+    appVersion: AppVersionEntry,
+    webappPackageVersion?: WebAppPackageVersionEntry,
+  ): Promise<Uint8Array> {
+    // For simplicity make all calls with one host. If that proves to not work well, split into
+    // separate calls to different hosts
+    return this.portalZomeClient.tryWithHosts<Uint8Array>({
+      fn: async (host) => {
+        // 1. get WebappPackageVersion and verify its hash - only necessary if the entry is not already
+        // provided as a function argument.
+        if (!webappPackageVersion) {
+          webappPackageVersion =
+            await this.portalZomeClient.customRemoteCall<WebAppPackageVersionEntry>({
+              host,
+              call: {
+                dna: appVersion.apphub_hrl.dna,
```

Comment:
Not the first time I am seeing the term `hrl`. What does it mean?

Reply:
It means "Holochain Resource Locator" and is essentially a combination of [DnaHash, AnyDhtHash] that addresses a piece of content in a holochain DHT.